### DISCUSSION TITLE
Dynamic Client choice + typescript

### DISCUSSION THREAD

**Discussion by MartinP-C** [TanStack/query#3605] [Q&A]
I have a component that can render data from multiple sources.
Each source is subtly different, but only ever contain the same type.

For example:
```tsx
const data: Bus[] | Car[] = [];
// data WILL NOT be (Bus | Car)[]
<Layer data={data} />
```

I'm using react-query to fetch the data. I have created a client hook for each data type.
```tsx
const useBusClient = (options) => {
  return useQueriesClient<Bus[]>('bus', arrayOfFetchFn);
}

const useCarClient = (options) => {
  return useQueriesClient<Car[]>('car', arrayOfFetchFn);
}
```

When I render the Layer component it is dynamic from a config, so I can't manually specify which query hook is used.
It needs to be dynamic based on the config:
```tsx
const Layer = ({ type }) => {
  const client = type === 'bus' ? useBusClient : useCarClient;

  const responses = client(options);
  // Expect responses to be UseQueryResult<BusResponse>[] | UseQueryResult<CarResponse>[]

  return responses.flatMap( (response)  => { 
    if (response.success && response.data) {
      // Return another component using response.data which should be Bus[] | Car [];
    }
  });

 return null;
};

const Layers = () => {
  const config = [
    { type: 'bus', ... },
    { type: 'car', ... }
  ];

  return (
    <>
      { config.map((layer) => {
        return <Layer type={layer.type} />
      }) }
    </>
  )
}
```

I'm getting Typescript errors on response.data.
It's ending up as an array of the union of possible types: `(Bus | Car)[]`
instead of (as I was expecting) a union of arrays of single type: `Bus[] | Car[]`

What is the "proper" way to dynamically select a query/queries client?
As in the example, I can't manually specify the query client because it's from a config (that will be supplied by an API, so that rules out using JSX inside the config).

Or is there some kind of bug from distributive unions?

I will try to create a code sandbox for this.

**Comment by karol-janik** [comment]
I do believe that the `result` type should be `UseQueryResult<Car[], Error> | UseQueryResult<Bus[], Error>`. I don't think that the `useQueriesClient` is way to go...  Maybe it's better to create custom hook for fetching data on top on`useQuery`, which gives you control with states like `isLoading` & `isError`. There is example code, that may help you:
```

type Person = {
  name: string;
  height: string;
  mass: string;
};

type Houses = {
  url: string;
  name: string;
  region: string;
  coatOfArms: string;
};

const fetchPeople = async (): Promise<Person[]> => {
  const response = await fetch("https://swapi.dev/api/people/1");
  return response.json();
};

const fetchHouses = async (): Promise<Houses[]> => {
  const response = await fetch("https://anapioficeandfire.com/api/houses");

  return  response.json();
};

const usePeople = (): UseQueryResult<Person[], Error> => {
  return useQuery<Person[], Error>(["PEOPLE"], fetchPeople );
};

const useHouses = (): UseQueryResult<Houses[], Error> => {
  return useQuery<Houses[], Error>(["HOUSES"], fetchHouses);
};

interface LayerProps {
  data: Houses[] | Person[] | undefined;
}
const Layer: React.FC<LayerProps> = ({ data }) => {
 {....}
};

interface Props {
  type: "house" | "person";
}
const MyComponent: React.FC<Props> = ({ type }) => {
  const query = type === "person" ? usePeople : useHouses;
  const result:
   UseQueryResult<Person[], Error>
    | UseQueryResult<Houses[], Error> = query();

  if (result.isLoading) {
    return <p>Loading...</p>;
  }
if(result.isError){
    return <p>Error!</p>
}
  return <Layer data={response.data} /> // Use map
};


export function App() {
  const client = new QueryClient();
  return (
    <QueryClientProvider client={client}>
      <div className="App">
        <MyComponent type="person" /> // Do the magic with map or sth :P 
      </div>
    </QueryClientProvider>
  );
}

```

**Reply by TkDodo** [reply]
yep, that's definitely the way to go. and it also works without type annotations, [here is a playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAChBOBnA9gOygXigbwLACgopUBDAWwgC4pFh4BLVAcwG4CiALCepj4a2g2ZtCUMiUSIBdRqwIBfEQXpkwyeMBxQAroggBFbQhBR5UAGbxkZKAHJ4EEgGNgAWgCOR+CFsECoSCgACWRdCERMHHYdeAAbaSE5UVIKBNkRIgcmejQ04WinZBJgAHlzAEF4MikaGXz8RT98QtRaCwhgJw44ZDBY6CwJEFQnKAAKAEpqGCsyej0AHjgkNABtAF0APkxtvFEWtodENVaBqBIAdxJ6TXMOrrGAIj5gMCkAenfEK7B6ADoACYQABu7xIv3ekF6-XeAEZHhMMlAHMBtPB0EcTno-gArFCoSYiRr4AgHW73DghMIRQaIYajSbTWbzCALKl6RAbbYYXYFNCHcJYs6Xa7kzocJ4vN6UT4kUi-ZDmehORyoAFKhx-QpkMEQjihDkIpSiFFo9DIwX8iC4-GEhTGsk6PQ9PpnSY7KImjpmp0GLwgJYIfEbAA0UAAovArPBNmNVo8YOGSjAADLhx7rMN3cUu-pQRH2pqOsLs8KRd08z2Zb3o32GYxsg3hUMRqPqWPxoIlACqAGVw72M1mKaXEAWGsbGMAEOZnNAZr0InsiAEqFBnk3HlAAD7ryArVCPImk-maACyIAAwtYThBUJosGNsFBV6YmYuJh7l1BHZ5jJFXwwIC9yDNAtwAfl9XNoGoEsm0QJFHSObRYgfKA-28O1RE+IhcIAPXApoiHocxxmQ1C-nmFMigBWRP2-atUVrWxqJIWjmD+TjbCReRohIsZyOASjEEjaMJgYi0mPQWxRPUXxRF40R+ME4SAEkAX6cTokYn1bHoDSIHkohFJ02sVIBYoSGiHDcKgAj7SAA) of your example, showing that result is properly inferred.
